### PR TITLE
Tracking how much memory is committed per object heap

### PR DESCRIPTION
**Pull request by cshung**
The overall goal of this PR (and more to come) is to provide a per-object-heap level hard limit.

In particular, this PR introduced a new field named `committed_by_oh` that has 4 entries corresponding to small object heap, large object heap, pinned object heap, and supporting data structure. The values are the amount of memory committed in bytes.

The key change here is passing the `oh` parameters to the various `virtual_*` methods. These tell the method which object heap this virtual operation is meant for. The various call site needs to provide that value. Most of the time, we know the call's purpose, by looking at the locals or the flags on the associated segment.

<del>The rest of the code is meant for making sure these passed values are correct. In particular, all the code that is under the `VERIFY_COMMITTED_BY_OH` definition is optional and is meant for testing only. The amount of memory committed for an object heap should match with the committed memory calculated from the associated segments of that object heap. However, it is not *always* true. In particular, by the time the `virtual_*` operation started until the segment objects are updated (or threaded), it is not true. The `unsaved_virtual_operation_count` is meant for making sure that we do not assert for identical value if we knew the `virtual_*` operation has already happened but the data structure is not yet updated. The `begin_virtual_operation` and `complete_virtual_operation` pair is meant to update that variable, and also provides a pairing of these operations by the character argument.</del>

<del>The testing was a bit more ambitious than it has to be. We only need to make sure all the passed `oh` parameter value (and hence the `committed_by_oh` value) is correct under the case for `heap_hard_limit`, but I expanded the check so that the values are true even if `heap_hard_limit` is not there. This is to avoid accidental misuse of those parameter values in the future. A slight twist is that if `VERIFY_COMMITTED_BY_OH` is not defined, the `committed_by_oh` is not maintained at all if `heap_hard_limit` is not on, this is meant for performance. I wanted to avoid taking the critical section unnecessarily.</del>

### CODE REVIEW

**Comment by VSadov** [comment]
I am wondering about the ultimate purpose of this accounting. - Is that for diagnostics or we expect users to actually specify separate individual limits per kind of allocation?

I.E. User can allow XX megs for small objects, YY megs for large, ZZ megs for pinned...

**Comment by cshung** [comment]
> I am wondering about the ultimate purpose of this accounting. - Is that for diagnostics or we expect users to actually specify separate individual limits per kind of allocation?
> 
> I.E. User can allow XX megs for small objects, YY megs for large, ZZ megs for pinned...

Yes. the overall goal of this PR (and more to come) is to provide (i.e. let user specifies) a per-object-heap level hard limit.

**Comment by Maoni0** [comment]
this verification code seems unnecessarily complicated. I would propose a much simpler way to verify. 

the reason why `virtual_commit` and `virtual_decommit` takes a global cs is because they need to care about the global use of the committed memory so we don't exceed the hardlimit. but for verification you can just verify per heap. and the methods that could modify a heap_segment's committed are already synchronized per heap and they are either PER_HEAP or take a heap number or heap. so you can keep `committed_by_oh_per_heap` and you don't need to take any locks because these operations are already guarded by the per heap lock when needed. you can get rid of the `begin/end_virtual_operation` and just put the verification each time after we update committed. delete segments is a none issue for hardlimit but if you really want to verify the non hardlimit scenario too, you can manually update the committed_by_oh_per_heap when we delete segments. 

you can verify that the committed_by_oh is the sum of `committed_by_oh_per_heap` of all heaps in a GC if you want. 

what do you think?

**Comment by cshung** [comment]
> what do you think?

I think that is a very interesting idea, this can hit 3 birds in one stone:

- Simplify the logic,
- Take `check_commit_cs` less frequently, and
- Verify (in addition to the `committed_by_oh` values) that the heap (number) is also correct.

The verification code was placed under `VERIFY_COMMITTED_BY_OH` for a reason. I knew they are non-trivial (and therefore could have bugs or complexity issue), they were designed to be optional. 

For the interest of time, I think we can proceed with this plan:

- Delete the code under `VERIFY_COMMITTED_BY_OH`
- Merge the PR
- File a tracking issue to do the testing work, and
- Complete the per object heap limit config switch feature.

I ran the stress tests (for both workstation and server GC, both without `heap_hard_limit`) with `VERIFY_COMMITTED_BY_OH` turned on. No assert was recorded for 24 hours each. Therefore I am pretty sure the code is correct.

**Comment by Maoni0** [comment]
I would suggest to not merge the PR as is. it's true that it's under an optional define but we should not check in code even under an optional define when we know we would change it and the change doesn't require that much effort (I think this should be much easier to implement than the one you did). I would estimate that since you now have looked through the relevant code paths it wouldn't take more than a few days, right?

**Comment by cshung** [comment]
> I would suggest to not merge the PR as is. it's true that it's under an optional define but we should not check in code even under an optional define when we know we would change it

I think you might have misunderstood what I just said. I have pushed the change to delete the code under the definition.

> it wouldn't take more than a few days, right?

I think I shouldn't take more than a week. I was just thinking it is probably a better idea to do the config switches first. Either way will work fine for me.

**Comment by mjsabby** [comment]
@cshung can you make sure Large Pages is tested in this case? even manually? I'll be able to put it on bing.com as soon as your build shows up in rolling builds.

**Comment by Maoni0** [comment]
sorry @cshung, I misread. I'm totally fine with checking the rest of the code but I have not looked at it in detail yet. let me do that. I see the description of the PR still mentions the code you deleted, BTW.

File: src/coreclr/src/gc/gc.cpp [hunk 1]
```diff
@@ -5569,41 +5566,58 @@ bool gc_heap::virtual_commit (void* address, size_t size, int h_number, bool* ha
                               virtual_alloc_commit_for_heap (address, size, h_number)) :
                               GCToOSInterface::VirtualCommit(address, size));
 
-    if (!commit_succeeded_p && heap_hard_limit)
+    if (!commit_succeeded_p)
```
Comment (Maoni0) [inline src/coreclr/src/gc/gc.cpp:5569]: >if (!commit_succeeded_p) [](start = 4, length = 24)

could revert this change since you don't have the code for verification

File: src/coreclr/src/gc/gc.cpp [hunk 2]
```diff
@@ -5569,41 +5566,58 @@ bool gc_heap::virtual_commit (void* address, size_t size, int h_number, bool* ha
...
+bool gc_heap::virtual_decommit (void* address, size_t size, gc_oh_num oh, int h_number)
 {
 #ifndef HOST_64BIT
     assert (heap_hard_limit == 0);
 #endif //!HOST_64BIT
 
     bool decommit_succeeded_p = GCToOSInterface::VirtualDecommit (address, size);
 
-    if (decommit_succeeded_p && heap_hard_limit)
+    if (decommit_succeeded_p)
```
Comment (Maoni0) [inline src/coreclr/src/gc/gc.cpp:5596]: same here

File: src/coreclr/src/gc/gc.cpp [hunk 3]
```diff
@@ -5569,41 +5566,58 @@ bool gc_heap::virtual_commit (void* address, size_t size, int h_number, bool* ha
...
+        }
     }
 
     return decommit_succeeded_p;
 }
 
+void gc_heap::virtual_free (void* add, size_t allocated_size, heap_segment* sg)
+{
+    assert(!heap_hard_limit);
+    GCToOSInterface::VirtualRelease (add, allocated_size);
```
Comment (Maoni0) [inline src/coreclr/src/gc/gc.cpp:5615]: >GCToOSInterface::VirtualRelease (add, allocated_size); [](start = 4, length = 54)

this could actually fail, could take this opportunity to fix it especially now you want to track things closely

File: src/coreclr/src/gc/gcpriv.h [hunk 4]
```diff
@@ -238,7 +239,7 @@ const int policy_expand  = 2;
 #ifdef SIMPLE_DPRINTF
 
 void GCLog (const char *fmt, ... );
-#define dprintf(l,x) {if ((l <= 1) || (l == GTC_LOG)) {GCLog x;}}
+#define dprintf(l,x) {if ((l == COMMIT_ACCOUNTING_LOG)) {GCLog x;}}
```
Comment (Maoni0) [inline src/coreclr/src/gc/gcpriv.h:242]: I would revert this for the checkin.

File: src/coreclr/src/gc/gcpriv.h [hunk 5]
```diff
@@ -398,6 +399,17 @@ enum gc_tuning_point
...
+enum gc_oh_num
+{
+    soh = 0,
+    loh = 1,
+    poh = 2,
+    none = 3,
+    total_oh_count = 4
+};
+
+gc_oh_num gen_to_oh(int gen);
```
Comment (Maoni0) [inline src/coreclr/src/gc/gcpriv.h:411]: nit, space between function name and '('

File: src/coreclr/src/gc/gcpriv.h [hunk 6]
```diff
@@ -4771,6 +4796,22 @@ BOOL heap_segment_uoh_p (heap_segment * inst)
     return !!(inst->flags & (heap_segment_flags_loh | heap_segment_flags_poh));
 }
 
+inline gc_oh_num heap_segment_oh (heap_segment * inst)
```
Comment (Maoni0) [inline src/coreclr/src/gc/gcpriv.h:4799]: seems to be unused now; if so please remove

File: src/coreclr/src/gc/gc.cpp [hunk 7]
```diff
@@ -38347,3 +38374,22 @@ void PopulateDacVars(GcDacVars *gcDacVars)
     UNREFERENCED_PARAMETER(gcDacVars);
 #endif // DACCESS_COMPILE
 }
+
+gc_oh_num gen_to_oh(int gen)
```
Comment (Maoni0) [inline src/coreclr/src/gc/gc.cpp:38378]: I would move this toward the beginning of the file as the later file is for GCHeap and other implementations

**Comment by cshung** [comment]
> @cshung can you make sure Large Pages is tested in this case? even manually? I'll be able to put it on bing.com as soon as your build shows up in rolling builds.

@mjsabby I am very happy that Bing is going to try it out. This is going to be a very good validation of the work.

This PR is indeed trying to solve the [3x commit problem](https://github.com/dotnet/runtime/pull/35802#issuecomment-624337471) we discovered earlier.

However, this PR alone is insufficient, as it only tracks the committed memory per object heap. In particular, it doesn't have the new configuration switches yet.

I am currently working on introducing the new switches, with that, I will alter the initial reservation logic, and add the per object heap commit limit checks. With that, we will solve the 3x commit problem.

I will make sure large pages are tested when I introduce the switches.

**Review by Maoni0 — APPROVED** [review]
LGTM!
